File: src/core/Cryptos/Events.cs
using System;
using core.Shared;

namespace core.Cryptos
{
    internal class CryptoDeleted : AggregateEvent
    {
        public CryptoDeleted(Guid id, Guid aggregateId, DateTimeOffset when) : base(id, aggregateId, when)
        {
        }
    }

    public class CryptoPurchased : AggregateEvent, ICryptoTransaction
    {
        public CryptoPurchased(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            Guid userId,
            string token,
            decimal quantity,
            decimal dollarAmountSpent,
            string notes)
            : base(id, aggregateId, when)
        {
            UserId = userId;
            Token = token;
            Quantity = quantity;
            DollarAmountSpent = dollarAmountSpent;
            Notes = notes;
        }

        public Guid UserId { get; }
        public string Token { get; }
        public decimal Quantity { get; }
        public decimal DollarAmountSpent { get; }
        public string Notes { get; }
        public decimal DollarAmount => DollarAmountSpent;
    }

    public class CryptoSold :
        AggregateEvent,
        ICryptoTransaction
    {
        public CryptoSold(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            Guid userId,
            string token,
            decimal quantity,
            decimal dollarAmountReceived,
            string notes)
            : base(id, aggregateId, when)
        {
            UserId = userId;
            Token = token;
            Quantity = quantity;
            DollarAmountReceived = dollarAmountReceived;
            Notes = notes;
        }

        public Guid UserId { get; }
        public string Token { get; }
        public decimal Quantity { get; }
        public decimal DollarAmountReceived { get; }
        public string Notes { get; }

        public decimal DollarAmount => DollarAmountReceived;
    }

    public class CryptoAwarded :
        AggregateEvent,
        ICryptoTransaction
    {
        public CryptoAwarded(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            Guid userId,
            string token,
            decimal quantity,
            decimal dollarAmountWorth,
            string notes)
            : base(id, aggregateId, when)
        {
            UserId = userId;
            Token = token;
            Quantity = quantity;
            DollarAmountWorth = dollarAmountWorth;
            Notes = notes;
        }

        public Guid UserId { get; }
        public string Token { get; }
        public decimal Quantity { get; }
        public decimal DollarAmountWorth { get; }
        public string Notes { get; }
        public decimal DollarAmount => DollarAmountWorth;
    }

    public class CryptoYielded :
        AggregateEvent,
        ICryptoTransaction
    {
        public CryptoYielded(
            Guid id,
            Guid aggregateId,
            DateTimeOffset when,
            Guid userId,
            string token,
            decimal quantity,
            decimal dollarAmountWorth,
            string notes)
            : base(id, aggregateId, when)
        {
            UserId = userId;
            Token = token;
            Quantity = quantity;
            DollarAmountWorth = dollarAmountWorth;
            Notes = notes;
        }

        public Guid UserId { get; }
        public string Token { get; }
        public decimal Quantity { get; }
        public decimal DollarAmountWorth { get; }
        public string Notes { get; }
        public decimal DollarAmount => DollarAmountWorth;
    }

    internal class CryptoTransactionDeleted : AggregateEvent
    {
        public CryptoTransactionDeleted(
            Guid id,
            Guid aggregateId,
            Guid transactionId,
            DateTimeOffset when) : base(id, aggregateId, when)
        {
            TransactionId = transactionId;
        }

        public Guid TransactionId { get; }
    }

    internal class CryptoObtained : AggregateEvent
    {
        public CryptoObtained(Guid id, Guid aggregateId, DateTimeOffset when, string token, Guid userId) : base(id, aggregateId, when)
        {
            Token = token;
            UserId = userId;
        }

        public string Token { get; }
        public Guid UserId { get; }
    }
}

File: src/core2.fs/Cryptos/Events.fs
namespace core.Cryptos

open System
open core.Shared

type ICryptoTransaction =
    abstract Id: Guid
    abstract When: DateTimeOffset
    abstract Quantity: decimal
    abstract DollarAmount: decimal
    abstract Token: string
    abstract Notes: string

[<AllowNullLiteral>]
type internal CryptoDeleted(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset) =
    inherit AggregateEvent(id, aggregateId, ``when``)

type CryptoPurchased(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, token: string, quantity: decimal, dollarAmountSpent: decimal, notes: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Token = token with get
    member val Quantity = quantity with get
    member val DollarAmountSpent = dollarAmountSpent with get
    member val Notes = notes with get
    member this.DollarAmount = this.DollarAmountSpent
    interface ICryptoTransaction with
        member this.Id = this.Id
        member this.When = this.When
        member this.Quantity = this.Quantity
        member this.DollarAmount = this.DollarAmount
        member this.Token = this.Token
        member this.Notes = this.Notes

type CryptoSold(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, token: string, quantity: decimal, dollarAmountReceived: decimal, notes: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Token = token with get
    member val Quantity = quantity with get
    member val DollarAmountReceived = dollarAmountReceived with get
    member val Notes = notes with get
    member this.DollarAmount = this.DollarAmountReceived
    interface ICryptoTransaction with
        member this.Id = this.Id
        member this.When = this.When
        member this.Quantity = this.Quantity
        member this.DollarAmount = this.DollarAmount
        member this.Token = this.Token
        member this.Notes = this.Notes

type CryptoAwarded(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, token: string, quantity: decimal, dollarAmountWorth: decimal, notes: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Token = token with get
    member val Quantity = quantity with get
    member val DollarAmountWorth = dollarAmountWorth with get
    member val Notes = notes with get
    member this.DollarAmount = this.DollarAmountWorth
    interface ICryptoTransaction with
        member this.Id = this.Id
        member this.When = this.When
        member this.Quantity = this.Quantity
        member this.DollarAmount = this.DollarAmount
        member this.Token = this.Token
        member this.Notes = this.Notes

type CryptoYielded(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, userId: Guid, token: string, quantity: decimal, dollarAmountWorth: decimal, notes: string) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val UserId = userId with get
    member val Token = token with get
    member val Quantity = quantity with get
    member val DollarAmountWorth = dollarAmountWorth with get
    member val Notes = notes with get
    member this.DollarAmount = this.DollarAmountWorth
    interface ICryptoTransaction with
        member this.Id = this.Id
        member this.When = this.When
        member this.Quantity = this.Quantity
        member this.DollarAmount = this.DollarAmount
        member this.Token = this.Token
        member this.Notes = this.Notes

[<AllowNullLiteral>]
type internal CryptoTransactionDeleted(id: Guid, aggregateId: Guid, transactionId: Guid, ``when``: DateTimeOffset) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val TransactionId = transactionId with get

[<AllowNullLiteral>]
type internal CryptoObtained(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset, token: string, userId: Guid) =
    inherit AggregateEvent(id, aggregateId, ``when``)
    member val Token = token with get
    member val UserId = userId with get
